### ISSUE TITLE
AArch64 byte/short stack saves and stack copies are broken on AArch64

### ISSUE THREAD

**Issue by adinn** [oracle/graal#236]
I have found an error in the implementation of byte/short stack saves and stack copies on AArch64. This error has been exacerbated by the change to "use scaled addressing in preference to unscaled where possible" (PR #233). However, the error was present before that patch went in.

I will attach a reproducer which manifests the error

- as a test failure before PR#233 was merged
- as a VM crash with PR #233 merged.

Analysis of problem:

The problem lies in the implementation of methods `AArch64Move.reg2stack()` and `AArch64Move.emitStackMove()`. They currently size the ldr/str operations which implement the necessary data transfer(s) as 32 bit operations:

```
    private static void reg2stack(CompilationResultBuilder crb, AArch64MacroAssembler masm, AllocatableValue result, AllocatableValue input) {
        AArch64Address dest = loadStackSlotAddress(crb, masm, asStackSlot(result), Value.ILLEGAL);
        Register src = asRegister(input);
        AArch64Kind kind = (AArch64Kind) input.getPlatformKind();
        final int size = kind.getSizeInBytes() * Byte.SIZE;
        if (kind.isInteger()) {
            masm.str(size, src, dest);
        } else {
            masm.fstr(size, src, dest);
        }
    }
    . . .
    private static void emitStackMove(CompilationResultBuilder crb, AArch64MacroAssembler masm, AllocatableValue result, Value input) {
        try (ScratchRegister r1 = masm.getScratchRegister()) {
            try (ScratchRegister r2 = masm.getScratchRegister()) {
                Register rscratch1 = r1.getRegister();
                Register rscratch2 = r2.getRegister();
                PlatformKind kind = input.getPlatformKind();
                final int size = kind.getSizeInBytes() <= 4 ? 32 : 64;

                // Always perform stack -> stack copies through integer registers
                crb.blockComment("[stack -> stack copy]");
                AArch64Address src = loadStackSlotAddress(crb, masm, asStackSlot(input), rscratch2);
                masm.ldr(size, rscratch1, src);
                AArch64Address dst = loadStackSlotAddress(crb, masm, asStackSlot(result), rscratch2);
                masm.str(size, rscratch1, dst);
            }
        }
    }
```
Note that the input to method `reg2stack()` is always a register of type `DWORD`.

The problem with this code is that the address operands passed to the ldr/str instructions are computed using `StackSlot` inputs allocated on the assumption that the datum written to/read from the slot will be of size 1 or 2 bytes for, respectively, a saved byte or short.

Prior to PR #233 stack writes employed an UNSCALED address which employed the offset appropriate to the slot type as its displacement. So, a save/copy employing a word write to a short slot would store the correct short value in the desired slot but also overwrite the next short slot. Similarly, a save/copy employing a word write to a byte slot would store the correct byte value in the desired slot but would also overwrite the next 3 byte slots. The only relevant existing unit test `StackMoveTest` does write/read adjacent short slots but, unfortunately, the second write does not invalidate the first. The new test `StackStoreLoadTest` ensures that one write invalidates another and that the invalid data is read back and checked.

After PR #233 all stack writes employ a SCALED address with the scaling computed appropriately for a long, int, short or byte write according to the slot size (8, 4, 2 or 1, respectively) and the displacement computed by dividing the supplied offset by that scaling. Note, however, that this same format could also have been employed prior to PR #233 in the case that the unscaled offset was too large to fit into a 9-bit signed field. This use of an address scaled for a short or byte slot with a store sized to 32 bits is a disaster waiting to happen and it does indeed happen.

If, say a short slot has unscaled offset 18 then the address will have scaling 2 and displacement 9. However, a generated ldrw instruction using that address will scale the displacement by 4, writing the stack slot at offset 36, possibly eclipsing some other slot or, worse, in the case of a larger offset, writing off the end of the stack itself. The error is not even pursued consistently. A ldrw generated from `emitStackMove` will load consistently using the same overscaled offset. Unfortunately, a ldrw generated from `stack2reg` will be scaled using the stack slot size. So, for the above hypothetical slot any attempt to read back in the saved short will try to load it using offset 18.

Solution:

The fix is to modify methods `reg2stack` and `emitStackMove` so that they compute the size using the `AArch64Kind` of the slot being written, ensuring that short values are stored and loaded using strh and ldrh and that byte values are stored and loaded using strb and ldrb.

**Comment by adinn** [comment]
Here is the relevant unit test

```
/*
 * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
 * Copyright (c) 2017, Red Hat Inc. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
package org.graalvm.compiler.lir.jtt;

import org.junit.Before;
import org.junit.Test;

import org.graalvm.compiler.core.common.LIRKind;
import org.graalvm.compiler.lir.Variable;
import org.graalvm.compiler.lir.VirtualStackSlot;
import org.graalvm.compiler.lir.framemap.FrameMapBuilder;
import org.graalvm.compiler.lir.gen.LIRGeneratorTool;

import jdk.vm.ci.meta.JavaKind;
import jdk.vm.ci.meta.PlatformKind;
import jdk.vm.ci.meta.Value;
import jdk.vm.ci.meta.ValueKind;

public class StackStoreLoadTest extends LIRTest {
    private static PlatformKind byteKind;
    private static PlatformKind shortKind;

    @Before
    public void setUp() {
        byteKind = getBackend().getTarget().arch.getPlatformKind(JavaKind.Byte);
        shortKind = getBackend().getTarget().arch.getPlatformKind(JavaKind.Short);
    }

    private static class StackStoreLoadSpec extends LIRTestSpecification {
        @Override
        public void generate(LIRGeneratorTool gen, Value a) {
            FrameMapBuilder frameMapBuilder = gen.getResult().getFrameMapBuilder();
            ValueKind<?> valueKind = getValueKind(a);

            // create slots
            VirtualStackSlot s1 = frameMapBuilder.allocateSpillSlot(valueKind);
            VirtualStackSlot s2 = frameMapBuilder.allocateSpillSlot(valueKind);
            VirtualStackSlot s3 = frameMapBuilder.allocateSpillSlot(valueKind);

            // start emit
            gen.emitMove(s1, a);
            gen.emitMove(s2, a);
            gen.emitMove(s3, a);
            gen.append(gen.getSpillMoveFactory().createStackMove(s1, s3));
            Variable result = gen.emitMove(s2);
            Value slot1 = gen.emitMove(s1);
            Value slot3 = gen.emitMove(s3);
            // end emit

            // set output and result
            setResult(result);
            setOutput("slot1", slot1);
            setOutput("slot3", slot3);
        }

        protected ValueKind<?> getValueKind(Value value) {
            return value.getValueKind();
        }
    }

    private static final LIRTestSpecification stackCopy = new StackStoreLoadSpec();

    /*
     * short
     */

    private static final LIRTestSpecification shortStackCopy = new StackStoreLoadSpec() {
        @Override
        protected ValueKind<?> getValueKind(Value value) {
            return LIRKind.value(shortKind);
        }
    };

    @SuppressWarnings("unused")
    @LIRIntrinsic
    public static short copyShort(LIRTestSpecification spec, short a) {
        return a;
    }

    public short[] testShort(short a, short[] out) {
        out[0] = copyShort(shortStackCopy, a);
        out[1] = getOutput(shortStackCopy, "slot1", a);
        out[2] = getOutput(shortStackCopy, "slot3", a);
        return out;
    }

    @Test
    public void runShort() throws Throwable {
        runTest("testShort", Short.MIN_VALUE, supply(() -> new short[3]));
        runTest("testShort", (short) -1, supply(() -> new short[3]));
        runTest("testShort", (short) 0, supply(() -> new short[3]));
        runTest("testShort", (short) 1, supply(() -> new short[3]));
        runTest("testShort", Short.MAX_VALUE, supply(() -> new short[3]));
    }

    /*
     * byte
     */

    private static final LIRTestSpecification byteStackCopy = new StackStoreLoadSpec() {
        @Override
        protected ValueKind<?> getValueKind(Value value) {
            return LIRKind.value(byteKind);
        }
    };

    @SuppressWarnings("unused")
    @LIRIntrinsic
    public static byte copyByte(LIRTestSpecification spec, byte a) {
        return a;
    }

    public byte[] testByte(byte a, byte[] out) {
        out[0] = copyByte(byteStackCopy, a);
        out[1] = getOutput(byteStackCopy, "slot1", a);
        out[2] = getOutput(byteStackCopy, "slot3", a);
        return out;
    }

    @Test
    public void runByte() throws Throwable {
        runTest("testByte", Byte.MIN_VALUE, supply(() -> new byte[3]));
        runTest("testByte", (byte) -1, supply(() -> new byte[3]));
        runTest("testByte", (byte) 0, supply(() -> new byte[3]));
        runTest("testByte", (byte) 1, supply(() -> new byte[3]));
        runTest("testByte", Byte.MAX_VALUE, supply(() -> new byte[3]));
    }
}
```

**Comment by tkrodriguez** [comment]
This appears to have been fixed by #237
